### PR TITLE
Allow to provide extra labels

### PR DESCRIPTION
**Pull request by felixfontein**
Sometimes you don't want to add extra collections to the docs build to resolve some labels, but just provide these label targets. This is useful for community.general which has redirects to infoblox.nios_modules, but we don't want to install that collection to build the docs (it's not under our control and might produce new build errors). Building without that collection gives warnings such as
```
.../rst/collections/community/general/nios_a_record_module.rst:21: WARNING: undefined label: ansible_collections.infoblox.nios_modules.nios_a_record_module
```
which would require us to be lenient on docs build errors. With this new option we can add a file (rst/_targets.rst) that contains the given labels.

(Not yet tested.)

### CODE REVIEW

**Comment by briantist** [comment]
This is a good idea, my initial reaction is that I'd maybe like to see this as part of the `build.sh`.

It dovetails with some of my thoughts about `--lenient` and `--fail-on-errors`, in that I'd like to be able to pass them (and maybe other things) to `build.sh`, optionally, so that `sphinx-init` isn't the only time when they can be chosen.

So it might fit well with that idea. That would also allow this extra-labels feature to be used when building locally (not using the GH actions).

It wouldn't be too bad to add it only to these actions for now, and do the above later, but would like to see tests added since [we're actually testing this action](https://github.com/ansible-community/github-docs-build/blob/main/.github/workflows/test-action-build-init.yml)! I realize it's still WIP.

**Comment by felixfontein** [comment]
I don't think this should be part of `build.sh`. This is mainly a hack for CI, but never something you should ever do in a real docsite.

File: actions/ansible-docs-build-init/action.yml
Comment (briantist) [inline actions/ansible-docs-build-init/action.yml:40]: Discussed a bit in IRC, but the typical way of accepting multi-valued input for actions is is  via multi-line string. So in invocation it'd be:

```yaml
- uses: ....
  with:
    provide-link-targets: |
      target1
      target2
```

We don't necessarily have to do that; it's just a typical convention.

For example, see the use of `restore-keys` in the cache action:
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action

Comment (felixfontein) [reply actions/ansible-docs-build-init/action.yml]: I rebased, and rewrote it to use newlines instead of commas.

(I'm a bit worried of using inputs in scripts though. If I'm writing `echo "${{ input.foo }}"`, and the user passes `"\nrm -rf /\necho "` as `foo`, would then the three commands `echo ""`, `rm -rf /`, `echo ""` be executed? I don't think this is a new problem with GHA in general though...)

Comment (briantist) [reply actions/ansible-docs-build-init/action.yml]: Possibly it could be mitigated by setting an env var first with a here-doc or something? Or maybe even with the `env:` block on the step (although I think that might be problematic). Though I don't know then if it's the same problem when using the variable. We could play with some possibilities. It'd also be fine to do this as a separate step if you want, in Python or whatever, if that's more comfortable.

The most important safety mechanism we have is that this will only ever run with a `contents: read` token, so the "damage" even if execution is possible, is probably limited to something like `exit 0` to ensure the step is successful when it shouldn't be.

**Comment by briantist** [comment]
We do have tests (for the action, but not the workflow yet), so I would like to see the tests updated to hit this parameter and do at least some naïve checking on the output. If mitigations are added for the malicious input case, we could probably also add a check for that.

File: actions/ansible-docs-build-init/action.yml
Comment (felixfontein) [reply actions/ansible-docs-build-init/action.yml]: I think the only way to properly mitigate this is to set an env var with GHA itself. Everything else has the same drawbacks, or makes everything more complicated. I've implemented it in 13b1c0558072f943448b5eb5b7dac10fb083e475.

**Comment by felixfontein** [comment]
Hmm, it looks like the tests do not use the actions from this branch, but from the `main` branch:
```
Warning: Unexpected input(s) 'provide-links-safely', valid inputs are ['dest-dir', 'collections', 'skip-init', 'fail-on-error', 'lenient', 'antsibull-docs-version', 'provide-link-targets']
```
(See for example the `init` step of https://github.com/ansible-community/github-docs-build/runs/6949802920?check_suite_focus=true)

**Comment by briantist** [comment]
> Hmm, it looks like the tests do not use the actions from this branch, but from the `main` branch:
> 
> ```
> Warning: Unexpected input(s) 'provide-links-safely', valid inputs are ['dest-dir', 'collections', 'skip-init', 'fail-on-error', 'lenient', 'antsibull-docs-version', 'provide-link-targets']
> ```
> 
> (See for example the `init` step of https://github.com/ansible-community/github-docs-build/runs/6949802920?check_suite_focus=true)

hm, that should not be the case, we're using the "local" form where the action is loaded from the files on the runner (so via checkout):
- https://github.com/ansible-community/github-docs-build/runs/6949802920?check_suite_focus=true#step:4:3

And the event is `pull_request` and not `pull_request_target` so the default checkout (and the one we're using) is the merge commit:
- https://github.com/ansible-community/github-docs-build/runs/6949802920?check_suite_focus=true#step:2:156

I'll see what I can find, but looking briefly I think it might just be a mismatch on the name of the option provided.

File: .github/workflows/test-action-build-init.yml
Comment (briantist) [inline .github/workflows/test-action-build-init.yml:74]:
```suggestion
          provide-link-targets: ${{ matrix.provide-links-safely }}
```

File: .github/workflows/test-action-build-init.yml
Comment (briantist) [inline .github/workflows/test-action-build-init.yml:39]:
```suggestion
          - ''
          - |
```

Need to also add this to the `name:` field to differentiate the jobs, but I can't do it until I get back to my dev environment because GH still doesn't allow suggestions on unchanged lines (line 17). I'd probably recommend it being a bool, so something like:
```
, link-targets=${{ matrix.provide-links-safely == '' }}
```

File: .github/workflows/test-action-build-init.yml
Comment (briantist) [inline .github/workflows/test-action-build-init.yml:117]:
```suggestion
          # short circuit if not provided
          ${{ matrix.provide-links-safely == '' }} || grep -- '^:orphan:$' rst/_targets.rst || exit 1
          ${{ matrix.provide-links-safely == '' }} || grep -- '^.. _outside_reference_1:$' rst/_targets.rst || exit 1
          ${{ matrix.provide-links-safely == '' }} || grep -- '^.. _outside_reference_2:$' rst/_targets.rst || exit 1
```

**Review by briantist — COMMENTED** [review]
So with the small change I committed they pass, I have a few more suggestions in this batch, and one I can't suggest directly in the UI, so I won't apply them yet.

Will leave it up to you if you want to keep the matrix name mismatched or make it match the param name.

File: .github/workflows/test-action-build-init.yml
Comment (felixfontein) [reply .github/workflows/test-action-build-init.yml]: Ok, **that** was a stupid one... Thanks for spotting this!

File: .github/workflows/test-action-build-init.yml
Comment (felixfontein) [reply .github/workflows/test-action-build-init.yml:117]: Doesn't it need to short-circuit on both `matrix.skip-init` and `matrix.provide-links-safely == ''`?

Comment (briantist) [reply .github/workflows/test-action-build-init.yml:117]: In theory, yes actually, I think we're always setting it to `''` when `skip-init` is true due to it being an include in the matrix maybe? I didn't look closely enough though, so I trust you on that. It'll be longer but probably better to do both.

File: .github/workflows/test-action-build-init.yml
Comment (felixfontein) [reply .github/workflows/test-action-build-init.yml:40]: Implemented in 8d376978132dbfe3c030303e2565332a23cba043 and 699d8f7ddc8c98b1d46199701429ac3fd032aa04. I've used the inverse condition.

File: .github/workflows/test-action-build-init.yml
Comment (felixfontein) [reply .github/workflows/test-action-build-init.yml]: Implemented in 8d376978132dbfe3c030303e2565332a23cba043.

**Review by briantist — APPROVED** [review]
Looks great, thank you for adding the tests!
I'll hold off on merging in case you had any other changes you wanted to make. If not then it's ready to go from my POV, feel free to merge :)

**Comment by felixfontein** [comment]
I think I'm done (with the first version), so let's merge. I'll probably try this out with community.general, and then I'll see whether something is missing ;-)

**Comment by briantist** [comment]
thanks for contributing @felixfontein !

**Pull request merged**
